adding ORDER BY to OVER() on IBMDB2

guidofaecke opened the pull request:
Finished up the missing ORDER BY within OVER() on IBMDB2.  
Might not be the best and most elegant approach, but that's how it works for us.
I am sure many IBM shops would appreciate it.

guidofaecke committed:
adding ORDER BY to OVER() on IBMDB2

Finished up the missing ORDER BY within the OVER() on IBMDB2 platform.
Might not be the best and most elegant solution, but that's how it works for us.

guidofaecke committed:
Merge pull request #1 from guidofaecke/guidofaecke-patch-1

adding ORDER BY to OVER() on IBMDB2

AwakendSleeper2016 commented:
This looks solid enough.  With luck, this will get merged in.  Without it, doctrine can't be used to its fullest in IBM/db2 shops.

guidofaecke committed:
moved additional functionality into private method

Some source cleanup...

Ocramius commented:
This lacks testing though

guidofaecke committed:
added test method testModifiesLimitQueryWithOrderBy()

Silly me, forgot the tests...

guidofaecke committed:
prevent "*" from breaking it

guidofaecke commented:
@Ocramius You caught the missing tests, didn't you 😄

morozov reviewed (commented):
@guidofaecke is it possible to create a functional test which describes the value/purpose of this change? (which I'm not questioning). The existing test does describe what the resulting queries should look like but doesn't tell why.

It would be also nice to have a more complex case covered with multiple columns and `ASC`|`DESC` involved. Otherwise, removing corresponding parts of the implementation will not make the new test fail.

guidofaecke committed:
missing substr caused undefined index

guidofaecke commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,44 @@ protected function getReservedKeywordsClass()
+        $orderByArray = [];
+
+        //determine if 'ORDER BY' is part of the query
+        $orderByPosition = strrpos($query, 'ORDER BY');
```
Good point... let me get that fixed

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/DB2PlatformTest.php:
```diff
@@ -377,6 +377,33 @@ public function testModifiesLimitQuery()
+
+        self::assertEquals(
+            'SELECT db22.* FROM (SELECT db21.*, ROW_NUMBER() OVER(ORDER BY id) AS DC_ROWNUM FROM (SELECT * FROM user) db21) db22 WHERE db22.DC_ROWNUM <= 10',
+            $this->_platform->modifyLimitQuery('SELECT * FROM user', 10)
```
Doesn't this case cover the same execution path as the previous one? The implementation checks `if ($offset > 0)`, so there's no difference between `0` and `NULL`. And from the SQL standpoint, no offset means "start from zero".

guidofaecke replied:
I copied form the test case above and added the order by

morozov replied:
I think this test shouldn't focus on testing of different variations of offset and limit since its logic doesn't depend on those, but instead, focus on edge cases of the `ORDER BY` expression.

guidofaecke replied:
Good point... Will see what I can come with.
Everyone else is more than welcome to chime in 😄

morozov replied:
Since the regexp matches the last `ORDER BY` expression in the string (as far as I understand, this is to ignore sub-queries), there should be a test covering a query with a sub-query.

guidofaecke replied:
The test will be there as soon as I cover more edge cases.

guidofaecke replied:
Did I mention I don't like "edge cases" 🤣 
Can someone introduce more parts to this class, so we can access certain items before the SQL string is build?

morozov replied:
Before we proceed with the parser, could you explain what exactly a non-empty `OVER()` is needed in this case? As far as I understand, if it's empty, then the row number assigned in the order specified by `ORDER BY` which is exactly what's needed for implementing `LIMIT, OFFSET`. Doesn't it work in all cases, or it works better with a non-empty `OVER()`.

guidofaecke replied:
On IBMDB2 everything is a little bit more complicated (very unfortunate). The row_number is assigned based on what's in OVER(). If OVER() is empty, the row_number is based on the relative-record-number of the table. The result of your regular order by would not look like what you are expecting if used with row_number.
If you need, as in most cases, the row_number assignment based on your order_by, you need the order in OVER() as well.
Either way, I agree, the current test-cases don't make much sense.

morozov replied:
@guidofaecke thanks for the clarification. With the above said and the fact that the row number in this query is used for emulating `LIMIT`, let's start with a test case in which a selected subset of the rows doesn't match the expectation.

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/DB2PlatformTest.php:
```diff
@@ -377,6 +377,33 @@ public function testModifiesLimitQuery()
+        self::assertEquals(
+            'SELECT * FROM user ORDER BY id',
+            $this->_platform->modifyLimitQuery('SELECT * FROM user ORDER BY id', null, null)
+        );
```
This test could use `@dataProvider` to make assertions independent on each other.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,46 @@ protected function getReservedKeywordsClass()
+                    case 'DESC':
+                        break;
+                    case '':
+                        $orderByArray[$orderIndex] = ',';
```
With this approach and the `preg_split()` expression above, if the `ORDER BY` columns have any delimiter than exactly ", ", there's a risk of getting malformed SQL. For example, if there's a leading white space(s) before the comma:
```php
var_dump(preg_split('/[, ]/', 'foo , bar , baz'));

/*
array(7) {
  [0] =>
  string(3) "foo"
  [1] =>
  string(0) ""
  [2] =>
  string(0) ""
  [3] =>
  string(3) "bar"
  [4] =>
  string(0) ""
  [5] =>
  string(0) ""
  [6] =>
  string(3) "baz"
}
*/
```
In this case, the resulting SQL will have more than one comma between the columns.

guidofaecke replied:
...and another fix I need to bang my head against

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,46 @@ protected function getReservedKeywordsClass()
+        $orderByPosition = strrpos($query, 'ORDER BY');
+
+        // ORDER BY found in query string
+        if (false !== $orderByPosition) {
```
If there's no `ORDER BY`, the method can use early return to avoid the extra indentation below.

guidofaecke committed:
changed 'order by' search

removed case-sensitivity from order by search and allow extra whitespace

guidofaecke committed:
fixed whitespace issue

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,64 @@ protected function getReservedKeywordsClass()
+     * 
+     * @param  string $query
+     * 
+     * @return array|string[]|mixed[]
```
This doesn't look usable. Why doesn't it always return a string?

guidofaecke replied:
Good question... Let me fix that...

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,64 @@ protected function getReservedKeywordsClass()
+                foreach ($splitOrder as $splitIndex => $splitValue) {
+                    switch (strtoupper($splitValue)) {
+                        case 'ORDER':
+                        case 'BY':
```
Why not take the leading `ORDER BY` out of the parsing? It's already captured by the regex in the beginning, so we only need to parse the actual set of columns.

Also, the `ASC` and `DESC` can be only in the end of the column, so instead of splitting the column in words, can we just `preg_replace()` the trailing `ASC`|`DESC` with an empty string?

I believe the closer the parsing logic to the actual SQL syntax is, the more reliable it is in edge cases.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +880,64 @@ protected function getReservedKeywordsClass()
+                        case 'ASC':
+                        case 'DESC':
+                            break;
+                        case '':
```
I think this can be improved. There shouldn't be an empty token. The fact that we capture an empty token means that the delimiter doesn't capture enough characters.

guidofaecke committed:
changed return type to string

guidofaecke committed:
removing ORDER BY from array

guidofaecke committed:
added a comment

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+        $query = preg_replace('/\s+/', ' ', $query);
+
+        //determine if 'ORDER BY' is part of the query
+        $orderByFound = preg_match('/(\bORDER\b)(\s*)(\bBY\b)(?!.*\b\1\b)/i', $query, $orderByPositionArray, PREG_OFFSET_CAPTURE);
```
I believe it's better to use `preg_match_all()` and use the last match. Otherwise, a query like `ORDER BY my_order` will be probably ignored.

morozov replied:
Aren't `\b`'s around `\s` redundant (`\b)(\s*)(\b`)? Could it be just `\s+`?

guidofaecke replied:
Just tying to address possible whitespaces between ORDER and BY

morozov replied:
You're right, `my_order` will not match `\bORDER\b`.

guidofaecke replied:
This one can stay? Just checking!

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+            $orderByPosition = $orderByPositionArray[0][1];
+
+            $queryArray = preg_split('/[, ]/', substr($query, 0, $orderByPosition -1));
+            $orderByArray = explode(',', substr($query, $orderByPosition + strlen('ORDER BY')));
```
`strlen('ORDER BY')` will probably corrupt the query if there's more than one space between `ORDER` and `BY` in the original query.

guidofaecke replied:
I think line 898 should take care of the whitespace, which should simplify the whole preg_split and explode madness.

All I tried to do is help with the IBM issue, but you are worth in code review than I am! btw. That's a compliment 😆 

Anyway, maybe with some help we can get than show on the road; Alan, Clark... you are more than welcome in joining me...

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+     * 
+     * @return string
+     */
+    private function getOrderByForOver($query)
```
Since the DBAL requires PHP ^7.1, we should use scalar type hints and return types where applicable.

guidofaecke replied:
I was hoping someone would say that!
Based on the current code, without checking composer.json, I wasn't sure if I should apply that.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+        $orderByFound = preg_match('/(\bORDER\b)(\s*)(\bBY\b)(?!.*\b\1\b)/i', $query, $orderByPositionArray, PREG_OFFSET_CAPTURE);
+
+        // ORDER BY found in query string
+        if (0 !== $orderByFound) {
```
Please use an early return.

guidofaecke replied:
My bad. 
I was coding by talking myself through it.
You are absolutely right.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+        $orderByArray = [];
+
+        // remove extra white space
+        $query = preg_replace('/\s+/', ' ', $query);
```
This can (or even should) be replaced with a proper `preg_split()` expression once the `ORDER BY` part is detected. No need to process the entire query.

guidofaecke replied:
In here I try to eliminate dead whitespace.
However, if it bugs you, I will reconsider.

guidofaecke replied:
Also, it simplifies the processing a little bit along the line.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+        if (0 !== $orderByFound) {
+            $orderByPosition = $orderByPositionArray[0][1];
+
+            $queryArray = preg_split('/[, ]/', substr($query, 0, $orderByPosition -1));
```
The delimiter should expect any number of any valid whitespaces before and after the comma.

guidofaecke replied:
...bag to the whitespace issue... That's why 898 is there for tight now.. Looks like I should tweak it a little.

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+                        default:
+                            $splitOrder[$splitIndex] = array_search($splitValue, $queryArray) === false
+                                                       ? $splitValue
+                                                       : $queryArray[array_search($splitValue, $queryArray)+2];
```
Could you pls add some tests explaining this logic? As I read it, the code tries to find the same column in `SELECT` and if fails, uses the expression as is, otherwise, uses something else (the `+ 2` part). The variable names `$splitOrder` (which is an array of strings) and `$splitValue` look confusing. Is it possible to rename it using some SQL terms?

Not sure if this code properly handles the case when a selected column has an alias (with or without the `AS` keyword) like `SELECT foo foo_alias, bar AS bar_alias`. Not sure what the expected result in this case is. This should be covered by a test too.

guidofaecke replied:
The logic... DB2 (on IBMi) is kinda strict on certain statements...
The reason for that, and I still not to address the possibility of subqueries here, is -> by default, if I use the entity and slap a limit/offset on it, it always adds a "as xxx" to it. within the "OVER()" you have to sort by the "AS" declaration otherwise DB2 can't handle it.
However!, I have to see how IBM reacts to your statement; without the "AS"...
Dang... I thought that would be an easy ride, like he persistent connection I added to ZF2 for IBMi... LOL

morozov commented on lib/Doctrine/DBAL/Platforms/DB2Platform.php:
```diff
@@ -878,4 +881,57 @@ protected function getReservedKeywordsClass()
+                            break;
+                    }
+                }
+                $orderByArray[$orderIndex] = array_filter($splitOrder);
```
What are the expected empty values in this case? Should `'0'` be filtered too? Looks like it can be fixed by a proper replacement of the `explode(' ')` above.

guidofaecke replied:
not 100% sure yet, the filtered values are primarily "" (empty string).
Might have to investigate more...

guidofaecke commented:
@morozov Are you in Vegas in October (ZendCon)? We should connect!

guidofaecke committed:
early return added

guidofaecke committed:
added typecast

morozov commented:
> Are you in Vegas in October (ZendCon)? We should connect!

Not really. Maybe some other conference next year.

guidofaecke committed:
closing parentheses missing

guidofaecke committed:
variable spelled wrong

Apparently I can't type today...

guidofaecke committed:
missing 'AS' caused wrong selection

guidofaecke committed:
changing tests

guidofaecke committed:
fixing existing tests

Trying to come up with more cases, however...
- native queries do not run through modifyLimitQuery
- everything else is based on entities and looks similar to the 4th test

guidofaecke committed:
prevent unknown index

guidofaecke committed:
mixed limit and offset in the test

guidofaecke committed:
added mixed sort case

tried to come up with more test cases, but they all turned out to look like the ones added.

guidofaecke committed:
Changed as requested

- removed preg_replace
- skip 'dctrn_ prepared columns

guidofaecke committed:
Prevent preg_split from returning false

guidofaecke commented:
Closed in favor of #3016
